core/src/main/java/jenkins/security/ResourceDomainRootAction.java

```diff
@@ -0,0 +1,278 @@
+import hudson.security.ACL;
+import hudson.security.ACLContext;
+import hudson.security.AccessControlled;
+import jenkins.model.Jenkins;
+import jenkins.util.SystemProperties;
+import org.acegisecurity.AccessDeniedException;
+import org.acegisecurity.Authentication;
+import org.apache.commons.codec.binary.Base64;
+import org.apache.commons.lang.ArrayUtils;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.*;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static java.time.Instant.*;
+import static java.time.temporal.ChronoUnit.MINUTES;
+
+/**
+ * Root action serving {@link DirectoryBrowserSupport} instances
+ * on random URLs to support resource URLs (second domain).
+ *
+ * @see ResourceDomainConfiguration
+ * @see ResourceDomainFilter
+ *
+ * @since TODO
+ */
+@Extension
+@Restricted(NoExternalUse.class)
+public class ResourceDomainRootAction implements UnprotectedRootAction {
+
+    private static final Logger LOGGER = Logger.getLogger(ResourceDomainRootAction.class.getName());
+
+    @CheckForNull
+    @Override
+    public String getIconFileName() {
+        return null;
+    }
+
+    @CheckForNull
+    @Override
+    public String getDisplayName() {
+        return null;
+    }
+
+    @CheckForNull
+    @Override
+    public String getUrlName() {
+        return "static-files";
+    }
+
+    public void doIndex(StaplerRequest req, StaplerResponse rsp) throws IOException {
+        if (!ResourceDomainConfiguration.isResourceRequest(req)) {
+            rsp.sendError(404, "Cannot handle requests to this URL unless on Jenkins resource URL.");
+        } else {
+            rsp.sendError(404, ResourceDomainFilter.ERROR_RESPONSE);
+        }
+    }
+
+    public Object getDynamic(String id, StaplerRequest req, StaplerResponse rsp) throws Exception {
+        if (!ResourceDomainConfiguration.isResourceRequest(req)) {
+            rsp.sendError(404, "Cannot handle requests to this URL unless on Jenkins resource URL.");
+            return null;
+        }
+
+        Token token = Token.decode(id);
+        if (token == null) {
+            rsp.sendError(404, ResourceDomainFilter.ERROR_RESPONSE);
+            return null;
+        }
+
+        String authenticationName = token.username;
+        String browserUrl = token.path;
+
+
+        Instant creationDate = ofEpochMilli(token.timestamp);
+        if (creationDate.plus(VALID_FOR_MINUTES, MINUTES).isAfter(now()) && creationDate.isBefore(now())) {
+            return new InternalResourceRequest(browserUrl, authenticationName);
+        }
+
+        // too old, so redirect to the real file first
+        return new Redirection(browserUrl);
```

Comment:
Could we not just `sendRedirect` now, rather than delegating to a separate object?